lager/lenses/tuple.hpp

```diff
@@ -0,0 +1,194 @@
+                             std::index_sequence<Seq...>)
+{
+    using fold_op::operator+;
+    return (LAGER_FWD(whole) + ... +
+            multiget<Seq>(LAGER_FWD(lenses), LAGER_FWD(parts)));
+}
+
+template <typename Whole, typename LensTuple, typename PartTuple>
+decltype(auto) set_fold(Whole&& whole, LensTuple&& lenses, PartTuple&& parts)
+{
+    return set_fold_impl(
+        LAGER_FWD(whole),
+        LAGER_FWD(lenses),
+        LAGER_FWD(parts),
+        std::make_index_sequence<
+            std::tuple_size_v<std::remove_reference_t<LensTuple>>>());
+}
+
+// note: has to be const & to avoid potential moved from UB
+template <class T, size_t... Seq>
+auto dup_impl(T const& x, std::index_sequence<Seq...>)
+{
+    return std::tuple<decltype(static_cast<void>(Seq), x)...>(
+        (static_cast<void>(Seq), x)...);
+}
+
+template <size_t N, class T>
+auto dup(T&& x)
+{
+    return dup_impl(LAGER_FWD(x), std::make_index_sequence<N>());
+}
+
+inline auto fobj_view = [](auto&&... args) { return view(LAGER_FWD(args)...); };
+inline auto fobj_set  = [](auto&&... args) { return set(LAGER_FWD(args)...); };
+
+// due to some obscure behavior on MSVC, this has to be implemented in a
+// manually defined function object
+template <size_t N>
+struct element_t : zug::detail::pipeable
+{
+    template <typename F>
+    auto operator()(F&& f) const
+    {
+        return [f = std::forward<F>(f)](auto&& whole) {
+            return f(std::get<N>(LAGER_FWD(whole)))([&](auto&& part) {
+                auto res         = LAGER_FWD(whole);
+                std::get<N>(res) = LAGER_FWD(part);
+                return res;
+            });
+        };
+    }
+};
+
+} // namespace detail
+
+//! @defgroup lenses
+//! @{
+
+/**
+ * Lens<W1, P1>, ..., Lens<Wn, Pn> -> Lens<(W1, ..., Wn), (P1, ..., Pn)>
+ *
+ * Note: remember you can use the identity function as the identity lens.
+ */
+template <typename... Lenses>
+auto zip(Lenses&&... lenses)
+{
+    return zug::comp(
+        [lens_tuple = std::make_tuple(LAGER_FWD(lenses)...)](auto&& f) {
+            return [&, f = LAGER_FWD(f)](auto&& whole_tuple) {
+                using Whole = std::decay_t<decltype(whole_tuple)>;
+                return f(detail::apply_zip<Whole>(
+                    detail::fobj_view, lens_tuple, LAGER_FWD(whole_tuple)))(
+                    [&](auto&& part_tuple) {
+                        return detail::apply_zip<Whole>(detail::fobj_set,
+                                                        lens_tuple,
+                                                        LAGER_FWD(whole_tuple),
+                                                        LAGER_FWD(part_tuple));
+                    });
+            };
+        });
+}
+
+/**
+ * Lens<W, P1>, ..., Lens<W, Pn> -> Lens<W, (P1, ..., Pn)>
+ *
+ * Note: parts MUST NOT OVERLAP. if they do the fold will
+ * overwrite sequential writes to the part. just don't do it.
+ */
+template <typename... Lenses>
+auto explode(Lenses&&... lenses)
```

Comment:
Ok... so the names `zip` and `explode`, did you borrow them from somewhere?

After reading the unit tests, intuitively I feel like I would call `explode` `zip` instead. And `zip`... maybe I would call `apply` or `each` or alike?

Reply:
So my train of thought was that I was zipping _lenses_ together to match the size of a tuple. `apply` would work too, though it refers really to the actual viewing/setting of the resulting lens.

`explode` was named as such because it gets you a bunch of parts for a single whole, instead of just one part as most lenses do. I don't think `zip` conveys that same meaning terribly well.

`explode` also makes it look dangerous, which it can be if you're not careful and give it lenses with overlapping parts.

Maybe we could call it `split` or `fan` or something like this? `zip` really doesn't make sense to me for this...

Reply:
Well, it depends on what direction you are looking from. You can also think that you are starting with a bunch of parts, and zipping them together into a single whole ;)

Reply:
yeah but that's also what my current `zip` does. `explode` has the benefit of conveying the fact that it expects lenses with the same single whole...

Reply:
Ok... so the other day I was writting a long response trying to argue for other naming and I eventually realized that my arguments supported... your current naming! I've also tried to look into the Haskell lens package w/o much resolution. So maybe... I'm happy to accept whatever naming is your favourite. It stills feels a bit weird this zipping when you look at it from the POV of the data that goes through (not the lenses themselves), but I don't have a better answer. Maybe just `explode` -> `fan` for the latter combinator just to avoid something too agressive? :p

Reply:
I'm ok with `fan`, it describes the shape of the lens combination pretty well.